### ISSUE TITLE
Support for first-class image definition

### ISSUE THREAD

**Issue by jhvhs** [helm/helm#7754]
## Problem statement
As of Helm v 3.0, the only support for specifying docker images, is a placeholder structure for a single image in the starter chart. The absence of formal image definitions poses the following problems for both the chart authors and the end users:

- It is difficult or even impossible to automatically extract all the image references from a given helm chart. This is important for security, scanning and license checking purposes, and affects both the end users and the chart authors.
- It is difficult for the end user to easily install an existing helm chart using images that have been relocated to a private registry. There is a large number of users who must run their workloads behind a firewall, in air-gapped environments, or are restricted to using pre-approved internal registries.
- It is unclear for the chart authors how to refer to multiple images (see the side note below)
- Those issues dramatically escalate with the use of sub-charts.

| Side note on hard-coded image refs |
|----|
| When using multiple images, the chart authors often hard-code them, as there is no clear structure to follow. About half of the charts in `helm/charts` use more than one image. About a third of those multi-image charts hard-code image references simply because it’s the easiest thing to do. And in half of those hard-coding cases, all but one image are hard-coded. At the same time, of all 342 community charts only 3 hard-code all of the images. It feels as if the single image key in the template is almost encouraging the authors to hard-code the rest of the images. |
||

See below conversations within the community for more context:

https://kubernetes.slack.com/archives/C0NH30761/p1559589151166400
https://kubernetes.slack.com/archives/C0NH30761/p1571850216043700
https://kubernetes.slack.com/archives/C0NH30761/p1571156640168300

## Proposal
1. Add an `images.yaml` file to the helm chart root level with the structure described below, and illustrated by the following example:

```yaml
images: # Main section
  proxy: # Image identifier, unique within a chart
    registry: gcr.io # optional, when unset either directly or via locations, defaults to `docker.io`
    namespace: a-chart # optional, when unset either directly of via locations, defaults to `library`
    name: proxy
    tag: v1.2.3 # optional, defaults to `latest` (not recommended)
    digest: "sha256:d76f67a6c7d6e8f76b98a76a7ef154ad36c68a8e87f6d89b" # optional
  web: # The shortest definition
    name: web
charts: # Optional overrides for the sub-chart images
  dashboard: # Subchart identifier
    images: # Has the same structure as above
      prometheus: # all fields under each image key are optional for subcharts
        namespace: prom
        name: prometheus
        tag: v2.17.2
        digest: "sha256:fec7063177b209eaa3216bab35991924fceb7ba19fe5ce763c3e5cb7e705b586"
      grafana:
        name: grafana
        tag: 6.7.1
        digest: "sha256:9d43bebe9f3a5d27647231e1fa760644fafcc09a1ad85eeae1df40c491841bd9"
  tracing:
    images:
      opencensus:
        tag: 0.1.11
locations: # optional list of locations - see examples below
- host: my.registry.net # optional
  namespace: my-project # optional
  applyTo: # a list of images to apply to. Use * for all images in the current chart, *.* for all images in all charts
  - web # apply to an image in the current chart
  - "*" # apply to all images in the current chart
  - dashboard.grafana # apply to an image in a subchart
  - tracing.* # apply to all images in a subchart
  - "*.*" # apply to all images in all charts
```

The definition of the sections in the `images.yaml` file are as follows:

- `images`

  This section is mandatory, and contains definitions for the chart's images. Each image must be defined under a unique key that will be used in templates for referring to that image. Each image entry consists of the following fields:
  - `registry` (*optional*) fqdn for the registry host. If it is not set via any mechanisms, defaults to `docker.io`.
  - `namespace` (*optional*) the portion of the image path between the registry host name and the image name. See  [Steve Lasker's blog post](https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/) for more details.
  If not set via any mechanisms, defaults to `library`.
  - `name` the name of the image
  - `tag` (*optional*) the image tag
  - `digest` (*optional*) the image digest

- `charts`

  This section is optional, and is used for overriding image locations in subcharts when necessary.

  Because this section will only work for subcharts that follow the same image referencing principles, the `helm` CLI should produce an error and refuse to run when it detects that unsupported charts are listed in this section.

  To define an override, use the chart name as a key under `charts`, and place an images section under it, including only the fields that need to be overridden, as in the example above.

- `locations`

  This section is also optional, and is used for setting image registry or namespace for multiple images at once. It has an array of `location`s as its entire body.

  Each `location` must have either a `registry` field or a `namespace` field, or both. It must also have an `applyTo` array with at least one element in it. Each element of the `applyTo` array refers to one or more images, as follows:
  - `<image-name>` refers to the image defined in the current chart under the `<image-name>` key
  - `"*"` refers to all the images in the current chart
  - `<chart-name>.<image-name>` - refers to the image defined under the `<image-name>` key for the `<chart-name>` subchart
  - `<chart-name>.*` - refers to all the images in the `<chart-name>` subchart
  - `"*.*"` - refers to all the images in all charts

Given that most of the fields above are optional, the smallest `images.yaml` would look like this:

```yaml
images:
  nginx:
    name: nginx
```

2. An overriding file can be provided via the `--images` flag to all template-related helm commands. E.g.

```bash
$ helm install some-chart --images my-images.yaml
```

The overriding file should follow the same structure as `images.yaml` above, but only has to include the fields to be overridden.

3. Add a built-in template helper function to refer to the image:
```yaml
...
containers:
- name: nginx
  image: {{ image "nginx" }}
...
```

the helper uses the information in the `images.yaml` in the chart, its containing charts (if they are feature-compliant), as well as the user-supplied image file (if provided)

4. Provide a `helm get images` command to provide a BoM for operators or scripts:
```yaml
$ helm inspect images linkerd2 --images=my-img.yaml -o yaml
- image: gcr.io/linkerd-io/proxy
  name: proxy
  chart: linkerd2
- image: gcr.io/linkerd-io/proxy-init
  name: proxyInit
  chart: linkerd2
- image: docker.io/omnition/opencensus-collector
  name: collector
  chart: tracing
- image: docker.io/jaegertracing/all-in-one
  name: jaeger
  chart: tracing
```

## Backwards compatibility
5. Add a feature flag, `.Template.ImageDefinitions`  for backwards compatibility, so that it will be possible to run the charts with earlier versions of helm CLI.
```yaml
...
containers:
- name: nginx
{{- if .Template.ImageDefinitions }}
image: {{ image "nginx" }}
{{- else }}
image: {{ .Values.images.nginx }}
{{- end }}
```

## Changes to the linter

In order for this approach to work with sub-charts, it is important that as many charts as possible are following it. The linter should detect when the image references in the chart templates are not using the `image` helper for image references, and issue a warning, directing the chart authors to the documentation on how to use the image references.
The linter behaviour will be enabled only when the `--strict-image-references` flag is set.

**Comment by bacongobbler** [comment]
Hi there. I had a few questions about this proposal.

>     * It is difficult or even impossible to automatically extract all the image references from a given helm chart. This is important for security, scanning and license checking purposes, and affects both the end users and the chart authors.

It should be possible to write a command that accepts the output of `helm template`, parse the manifests for all workload types and spit out the images used in that chart. That should work today without requiring any introduction of new code. Have you considered going down that route first?

>     * It is unclear for the chart authors how to refer to multiple images (see the side note below)

> When using multiple images, the chart authors often hard-code them, as there is no clear structure to follow.

Curious. I thought there was a convention provided in the documentation. Upon further inspection, it looks like it was never documented.

A common convention for multiple images should look like the following:

```
images:
  nginx:
    repository: nginx
    pullPolicy: IfNotPresent
  busybox:
    repository: busybox
    pullPolicy: IfNotPresent
```

This could be rectified by providing guidelines for chart authors in the documentation, and the charts upstream could be changed to reflect this convention.

>     * It is difficult for the end user to easily install an existing helm chart using images that have been relocated to a private registry. There is a large number of users who must run their workloads behind a firewall, in air-gapped environments, or are restricted to using pre-approved internal registries.

Given the above convention, how is the current process made difficult? How does this proposal change that?

Finally, have you considered looking at [Schema files](https://helm.sh/docs/topics/charts/#schema-files) to provide this structure in your case?

I am concerned this proposal aims to provide a completely optional piece of code without any actual value gained in terms of Helm functionality. I'm not sure how this will solve issues for users looking for solutions with charts in the charts repo today. Looking at the use case more closely, I believe that the same use cases can be solved by providing new tooling that doesn't currently exist today.

**Comment by bacongobbler** [comment]
Can we please close either this ticket, #7332 or #7154? That way we can keep the conversation in one ticket. Thanks.

**Comment by jhvhs** [comment]
> It should be possible to write a command that accepts the output of `helm template`, parse the manifests for all workload types and spit out the images used in that chart. That should work today without requiring any introduction of new code. Have you considered going down that route first?

Yes, we have. Many charts contain conditionals around including some of the images, therefore, any given `helm template` output is never guaranteed to include all of the images.

> A common convention for multiple images should look like the following:

That convention is only followed by 2 out of 169 multi-image charts in `helm/charts` (namely, `stable/drone` and `stable/keel`). Hence, the original proposal #7154 .

> Given the above convention, how is the current process made difficult? How does this proposal change that?

The #7154 goes beyond that, and tackles the issue with sub-charts, setting the common properties, i.e. the registry host and/or namespace globally once for a given installation. The current proposal provides a more robust structure and removes a need for maintaining a heap of template helper boilerplate in every chart.

> I am concerned this proposal aims to provide a completely optional piece of code without any actual value gained in terms of Helm functionality
Image relocation is a core problem in containerised software packaging. And while it does not affect every user, it does affect a large portion of the community as more and more workloads are being run with on-premise deployments of Kubernetes.


> Can we please keep only one ticket open (either this ticket, #7332 or #7154)? 
#7332 is the interim proposal outline, while #7154 is a PR implementing it. This proposal is for a permanent solution. I do understand that they are related, but it would be nice to determine which approach to take before we close any of them. What would you recommend?

**Comment by technosophos** [comment]
_tl;dr: I am generally in favor of the proposal, but would like to discuss some changes. Most importantly, I would prefer a dedicated `images.yaml` file to changing Chart.yaml. This provokes several design changes._

This is probably going to be long. I took a look through the Helm code these last few days to see how much of this we could support. I also spent some time on CNAB and related tools to see about those. (Some of your colleagues helped us solve this exact problem in CNAB)

Two up-front statements on how I am viewing this:

1. This is not a frequently requested change. Nor has this particular proposal seen much community endorsement. So the conclusion I have to draw is that the status quo is satisfactory to the 80% case. I am not interested in doing anything that would alienate the majority, or make their life harder.
2. I am adamant that such a feature be initially released behind a feature flag. This is a disruptive change, and its ramifications are not well understood. It would be irresponsible to make this the "new default." I want the flexibility to experiment and see if this is going to work (and to have the flexibility to change it if it is not working.)

## Summary of my proposed changes

1. Use `images.yaml` instead of modifying `Chart.yaml`
2. Use an `--images=myimages.yaml` instead of modifying Values
3. Introduce an `image $imageName` template function (or `.Images.get($imageName)`) instead of `.Chart.Images.SomeImage.Reference`
4. Simplify the syntax to favor the 80% case instead of the %20 case (No globals, no default overrides, opt-in to inheritance rather than opt-out of inheritance)
5. Hide behavior behind a feature flag
6. Add `helm get images` and `helm inspect images` to commands

# 1. Altering the Chart.yaml file

Of all of our files, `Chart.yaml` is the hardest to modify without API breaking changes. When the Dependencies proposal came in for Helm 2, we introduced `requirements.yaml` to sidestep the compatibility issue until Helm 3, when we rolled that all into `Chart.yaml`.

I am also not convinced that adding images to `Chart.yaml` is correct, since the `Chart.yaml` does not carry any other parameterization content. A Chart.yaml is 100% true regardless of how it was installed. In your proposal, a Chart.yaml would have data that might not be true for the given installatiion.

The alternative would be to introduce an `images.yaml` which just contained the data you propose. That might require us to add a top-level `.Images` template object parallel to `.Chart`. I have some  other ideas later on in this response.

# 2. Overriding values

I am not a fan of using values to magically override things from `Chart.yaml` (or `images.yaml`). I can't tell for sure if that is the suggestion, but it seems to be. Values are intentionally free-form, and we already loath our decision to ever introduce `globals` as a special key. Mixing schematized and non-schematized content is irksome and error prone to implement, and is often confusing to users.

I would prefer instead to go the other route you suggested, and allow the user to specify an `images.yaml` override, and have that format be strongly schematized.

In that case, I would imagine the user would do something like:

```console
$ helm install foo bar --images=myimages.yaml --values=myvalues.yaml
```

I am not entirely convinced that you need global overrides for `imageRegistry` and `imageNamespace`. I think (and I could be convinced otherwise) that YAML references could solve this problem up front, and save a lot of work for the rest of the spec, since then `useOriginalNamespace`, etc would not be necessary.

When it comes to subcharts, in a strongly schematized format, we might prefer something like:

```yaml
# Supply values for a specific subchart
charts:
    subchart1:
      imageRegistry: default.reg.local # for all images in the sub-chart
      images:
          - name: example/nginx
            tag: v9.1
      charts:
          subsubchart:
              imageRegistry: ...
```

That is slightly more verbose, but I think it is also much clearer. I wish we'd done something closer to that with values.yaml. (Changing `charts` to `subchart` would be fine by me, too. I proposed `charts` since that is the name of the directory in a chart.)

# 3. Special Accessor

I'm really struggling with this part. I feel like the proposed method is too verbose, and I found it to be confusing for a few reasons.

Your examples add up to be a little unclear, so I wanted to try to understand something first. In your YAML, you have:

```yaml
images:
  name1:
    # ...
```

In your "special accessors" section, you have: `.Chart.Images.Nginx.`. Is the suggestion that we would rewrite the name the user supplied? In general, I don't like to automatically change a name that the user supplies. It's one thing to change reserved words, but another to change user data. You don't cover the more complex casing questions, like whether `foo_bar` gets changed to `Foo_bar` or `FooBar`. Do we disallow the user from using capital letters in `images`? If not, what do we do when `Name1` and `name1` are both set as keys? And what do we do to reverse-lookup `.Chart.Images.Name1`? (This second part is a problem even when there are not two keys.)

But the verbosity of what we expect the chart dev to do is a bigger concern. I feel like `.Chart.Images.Nginx.Reference` is too long of a qualified name that requires too much knowledge of an underlying data structure and the automatic `Reference` field/function. I'd almost rather use template functions to resolve all of this outside of the user's view. Someting like `{{ image . "nginx" }}` (returning `example.com/foo/nginx:v1.2.3`) feels far less cumbersome than building an image reference by hand or using a really long name.

That said, there probably is value in allowing access to each component part of the image. But that seems like more of an "expert mode" thing. We could do that either with another template function (`image_object . "nginx"`) or with the longer object format you provided (or by hanging accessors off if the `.Chart` object, e.g. `.Images.Get("foo").Tag`). There is an established pattern for this in the way we handle files in charts.

One advantage to using a template function: If we have to move the `images:` stuff behind the scenes, say from `images.yaml` to `Chart.yaml`, nothing will break and we won't have any hacky workarounds.

I have been trying to think through the potential pitfalls of using a template function with a context (`{{image . "foo"}}` vs `{{ image "foo" }}`). Contexts can be a little confusing, but I don't think we introduce any new complications by using them here. For example, a library chart would still have to take a context to access the images in the chart it is doing the processing for, regardless of which way we did this. For better or worse, chart devs are used to working with contexts.

# Linter

I would be willing to entertain adding linter rules as we got closer to Helm 4 or behind a feature flag. But I don't feel that this is something we need from the outset. I know that there is a segment of the population that would want to move to this right away, I am not sure this counts as an 80% use case at the outset.

Personally, I would prefer to initially hide this behind and "experimental" feature flag (which gives us the ability to change APIs related specifically to this feature) and see how closely this solution meets people's needs. See the OCI registry feature for an example of how we used this.

# Backward (and Forward) Compatibility

I think we may be able to do this in two phases: a non-breaking Helm 3 change, and a breaking Helm 4 change. The non-breaking Helm 3 version would allow this as a pattern, but not enforce it, while Helm 4 would require that all images be referenced.

I agree that the approach here _should_ be implementable in a backward compatible way, though the API for the Helm 3 version might be ugly as we try to implement it without breaking the existing API. So I imagine we will do a cleanup pass to the internals for Helm 4.

I am a uncertain about the potential effect on the Helm charts ecosystem. This will be a substantial break from Helm 2 charts, which might lead to a slower initial uptake. I would hope that we don't see a strong bifurcation of charts over this matter.

It appears that once we implement this, Helm users from earlier versions of 3.x will not be able to install charts that use these new features. That is probably okay, but it would not be bad to devise some patterns to make this transition less painful. One thing we're learning about enterprise software: you just can't assume people will upgrade quickly, even for minor releases.

# --reuse-values/--reset-values

One HUGE boon to using an `images.yaml` is that we won't have to fit this pattern, which is good. I don't think image management resembles the other forms of parameter management.

I'm leaning toward saying that if an `images.yaml` file is supplied, it is stored and re-used until another `image.yaml` is supplied, in which case the original is completely replaced.

# Referencing image by digest

Yes, this is a well known design flaw in OCI registies. I think we agree, though, that if a SHA is supplied, the SHA must be requested. That unfortunately requires that we provide a way to "zero out" the digest field. I think using the empty string is sufficient for that.

```yaml
images:
  name1:
    registry: special.reg.local
    name: myorg/my-product
    tag: "1.2.3"
    digest: ""
```

# Other Questions and Proposed Changes

## No `namespace`

Your concept of namespace does not match the OCI Distribution specification. https://github.com/opencontainers/distribution-spec/blob/master/spec.md#overview It _explicitly_ says that the things you call `name` and `namespace` are part of the same field (which the spec calls variously `name` or `repository`), and that there is no limit to only one path component. (I believe ACR and possibly other container registries support this.) While I am all for user-oriented simplifications, in this case I think it runs contrary to the specification.

## Subcharts and "Globals"

Your primary use case seems to be the case where users move ALL of their images to ONE registry. I am not sure that that is a frequent pattern. In neither Helm nor CNAB have I observed this as anything but an edge case (air gap scenarios and a few large enterprises that require on-prem image registries). In the vast majority of cases, only a few images ever get relocated. The examples are usually:

- A user forks and moves from one Docker Hub repo to another (`your/nginx` to `my/nginx`)
- A user relocates a few "internal" images (forks or specific code) but mostly uses off-the-shelf images from their upstream
- A user switches between dev and prod images of some set of images, which may entail moving between registries, or merely using two different repositories.

But the design as it stands seems to _favor_ storing all images in the same registry, requiring users to make exceptions for images that are not. I would rather see that design inverted.

For example, say I want `nginx:latest`. That is the _full name_ for the image from DockerHub (the most frequent source of images for Helm charts). Consider an example in your format:

```yaml
imageRegistry: default.reg.local # only applies to the main chart
imageNamespace: chart-owner # only applies to the main chart
images:
  myimage:
    name: foo
    tag: bar
  nginx:
    name: nginx
    tag: latest
```

In my reading of the proposal, the `nginx` image would resolve to `default.reg.local/chart-owner/nginx:latest`. Your path to remediation is:

```yaml
imageRegistry: default.reg.local # only applies to the main chart
imageNamespace: chart-owner # only applies to the main chart
images:
  myimage:
    name: foo
    tag: bar
  nginx:
    name: nginx
    tag: latest
    imageRegistry: ""
    imageNamespace: ""
```

That means the _average developer_ (probably near 80-90% of Helm chart developers) have to do _more_ work to get the base Docker image so that a small percentage of users don't have to re-type a registry name on repeat entries. Now if you work through that same case for overriding the original images, you'll see that it can get even more error-prone, as a user might try `useOriginalNamespace` or `useOriginalRegistry` only to discover that the upstream chart defines or redefines that.

The behavior of global image registries is particularly dangerous in this case. Thinking they are saving a few keystrokes in their local chart, developers are likely to accidentally set a global and then have broken deployments when their subcharts can't pull images. Imagine, for a moment, what a failed `helm upgrade` would look like in that case.

One potential workaround, which introduces more work for the < 20% use case, but is less work for the 80%, would be to make registries a namable reference like images are.

```yaml
registries:
    foo: foo.azurecr.io
    
images:
    myimage:
        name: foo/bar/baz
        tag: v1.2.3
        in: foo
    nginx:
        name: nginx
        tag: latest
```

In the above:
- `myimage` resolves to `foo.azurecr.io/foo/bar/baz:v1.2.3`
- `nginx` resolves to `nginx:latest`.

Then a registry can be overridden when necessary, and change all of the images in that registry. Similarly, an `in` field can be added in a `images.yaml` override to add a registry to `nginx` _if you want it_.

## Exposing images in Helm commands

The original proposal doesn't say anything about how I would find out what images a particular release is running. The mixture of Chart.yaml and values actually makes this quite hard for the user to discern.

I would imagine we'd want a `helm get images` command. And probably a `helm inspect images` command as well. Are there others we would need?

## Credentials

If we are getting this deep into image management, we should probably also ask whether we need to address credential management to any degree. In general, I am opposed to having Helm do credential management, but it is worth discussing.

**Comment by technosophos** [comment]
I am eager to let people start developing on this. Was anyone else still interested, or is this dead?

For my part, I will revisit after 3.3.0 work is done.

**Comment by TBBle** [comment]
I'm vaguely interested in this, and @technosophos 's write-up looks good.

We're currently following a consistent pattern of `image` with subkeys for `registry`, `repository`, and `tag`, and multi-images are `XImage` for the non-primary container image.

Since we're building a private system, we have a global `projectImageRegistry` value which all our charts honour to replace the `registry` for all our images with one on a different server, e.g., when deploying to production hardware which needs to use a mirror of our Docker registry.

One complication I'd like to inject is the use of skaffold, which currently expects to have a .Value it can override with an image reference for the image it built and pushed, in one of three structures:
* literal string value; or
* `repository` and `tag` sub-keys; or
* `registry`, `repository`, and `tag` sub-keys.

The latter option is the layout all our charts use. These three cover most of the charts I've looked at, but there's exceptions I've seen over time, and have mentioned in issue comments both here and on skaffold's issue tracker.

skaffold integration would be simplified given a structured _images.yaml_, but we'd have to get that support _into_ skaffold (which has been a bit slow about supporting Helm features...).

**Comment by jhvhs** [comment]
I'm really sorry for the long silence, these have been quite a few hectic weeks.

Let me walk through the changes suggested by @technosophos:

### Main points

- Introduce `images.yaml` rather than modifying `Chart.yaml` / Suggestions for syntax change

    I'm completely on board with `images.yaml`, especially given both the history of `requirements.yaml` and the mutability of the settings.
    In my opinion, in response to your comments on the override syntax and globals abuse, this is how `images.yaml` should look like:

    ```helmyaml
    locations:
    - host: gcr.io
      namespace: linkerd-io
      applyTo:
      - *
      - dashboard.grafana
    images:
      proxy:
        name: proxy
        tag: foo.bar.baz
      web:
        name: web
        tag: blah
      debug:
        name: debug
    charts:
      dashboard:
        images:
          prometheus:
            namespace: prom
            name: prometheus
            tag: v2.17.2
            digest: sha256:fec7063177b209eaa3216bab35991924fceb7ba19fe5ce763c3e5cb7e705b586
          metrics:
            namespace: deislabs
            name: smi-metrics
            tag: v0.2.1
          grafana:
            name: grafana # registry and namespace defined in `locations` block above
            tag: 6.7.1
            digest: sha256:9d43bebe9f3a5d27647231e1fa760644fafcc09a1ad85eeae1df40c491841bd9
      tracing:
        images:
          opencensus:
            namespace: omnition
            name: opencensus-collector
            tag: 0.1.11
    ```
    
    The difference between this version and the one above is how group overrides are handled. The `registry` and `namespace` can still be specified on a single image level, however, the group overrides will work by placing objects in the `locations` array. The structure must have either `host` or `namespace` set, or both, then the `applyTo` array will list images using the following rules:

    - images from the current chart are specified by their key, e.g. `debug`
    - images from subcharts have to be prepended by the subchart name and a dot, e.g. `tracing.opencensus`. If the subchart is nested, prepend it with a dot-separated list of the charts, e.g. `subchart1.subchart23.imagename`
    - all images in the current chart can be specified using a wildcard `*`
    - all images in a subchart can be specified using a wildcard prepended by the subchart name, e.g. `dashboard.*`
    - all images in all charts can be specified as `*.*`

    I believe this approach is as flexible as the original one, at the same time it does not require additional work, such as clearing the overrides. It also eliminates the need to fiddle with `globals` and `values` in general.

    The user can provide a file following the exact same structure, to override the values in the chart-provided file, where the locations will be internally appended to the array, and the last matching rule will win during rendering.

    The default values for registry host `docker.io` and registry namespace `library` are still assumed if nothing is specified.

- Change the accessor from `Charts.Images` to something more suitable

    Given that we agree to move the data out of `Chart.yaml` it makes sense to move the accessor out as well. Our options so far are:

    - `.Images.image_name` seems to be more strict on the image name, but less verbose
    - `.Images.get("image_name")` is more verbose, but allows for more relaxed image names. It is possible also to refer to images from subcharts, e.g.: `Images.get("my_subchart.my_image")`
    - A builtin template helper function `{{ image "image_name" }}` allows the same flexibility as above, but looks more concise. I'd lean towards that one.

- Hide this behaviour, and the associated commands behind a feature flag until v4.
    - If the functionality is implemented in any of the ways described above, installing the chart would fail when the flag is not set.
      I think it would be useful instead to have an internal flag the authors could check in the template in order to use the new functionality, and fallback to old methods otherwise, e.g.:
      ```helmyaml
      {{- if .Images -}}
      image: {{ image "nginx" }}
      {{- else -}}
      image: {{ .Values.image }}
      {{- end -}}
      ```
      In that case, regular charts would continue to work as is, and the charts with the new functionality will have a fallback.
    - I also like the idea of adding a special flag to the linter to ensure the new feature is used.

- Adding the `helm get images` and `helm inspect images` subcommands
    - I don't quite get what should the `helm get images` do. It does not make sense to download images locally. Does it make sense to download them as tarballs?
    - I have included a quick prototype in my presentation of `helm inspect images`, and have left it out of the list above. An example output would look like this:
    ```
      $ helm inspect images linkerd2 --images=my-img.yaml -o yaml
      - image: gcr.io/linkerd-io/proxy
        name: proxy
        chart: linkerd2
      - image: gcr.io/linkerd-io/proxy-init
        name: proxyInit
        chart: linkerd2
      - image: docker.io/omnition/opencensus-collector
        name: collector
        chart: tracing
      - image: docker.io/jaegertracing/all-in-one
        name: jaeger
        chart: tracing
    ```

- Handling the digest. I believe there are several ways to deal with a relocated digest. First of all, it seems more responsible for the user to actually retrieve the known digest for the image in the new registry and plug it in, in case it is different from the original one. That being said, I don't object against setting the value to an empty string - if someone wants to ignore it on purpose, let them.

### Other Concerns
- Namespace isn't defined by the OCI spec.
    
    Sure, it did not get its own separate name, but as you mention above, as the first use case, is `your/nginx` => `my/nginx` which is exactly that - a switch of a namespace. It's very unfortunate, that people can't refer to a specific part that has a meaning by name, and there were [some recent suggestions](https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/), among others, to use the term `namespace`.

- Moving **ALL** images to **ONE** registry is not a frequent pattern.
    
    It is the most frequent pattern for large corporate users - in some cases only the registry needs to be changed, in other cases both the registry and namespace have to be changed. In addition to that, Amazon is using opensource charts with images that they host, overriding the registry is the way to run charts in the [Amazon Marketplace](https://docs.bitnami.com/aws/get-started-charts-eks-marketplace/). Also seems to be the most common case for @TBBle. It seems that to support these use cases, we would want a pair of command-line switches, e.g. `helm install --global-registry-host=my.registry.local --global-registry-namespace=project23`.

- Credential management

    This is an interesting subject. I believe the credentials themselves are better off stored in the cluster, with helm only injecting the relevant secret names. However, I have come across use cases, when the developers wanted to create a namespace as part of `helm install`, and would want helm to provide a mechanism for generating and storing the credentials within the newly created namespace. As interesting as this discussion might turn out, I believe we should leave it for the next iteration.

**Comment by glyn** [comment]
> * Handling the digest. I believe there are several ways to deal with a relocated digest. First of all, it seems more responsible for the user to actually retrieve the known digest for the image in the new registry and plug it in, in case it is different from the original one. That being said, I don't object against setting the value to an empty string - if someone wants to ignore it on purpose, let them.

Relocating an image should preserve the digest so that the user can be sure that they are executing the same bits that were originally tested. See https://github.com/pivotal/image-relocation for a library for doing just that (or use the underlying ggcr library mentioned under "Alternatives" in the README).

> 
> ### Other Concerns
> * Namespace isn't defined by the OCI spec.
>   Sure, it did not get its own separate name, but as you mention above, as the first use case, is `your/nginx` => `my/nginx` which is exactly that - a switch of a namespace. It's very unfortunate, that people can't refer to a specific part that has a meaning by name, and there were [some recent suggestions](https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/), among others, to use the term `namespace`.
> * Moving **ALL** images to **ONE** registry is not a frequent pattern.
>   It is the most frequent pattern for large corporate users - in some cases only the registry needs to be changed, in other cases both the registry and namespace have to be changed. In addition to that, Amazon is using opensource charts with images that they host, overriding the registry is the way to run charts in the [Amazon Marketplace](https://docs.bitnami.com/aws/get-started-charts-eks-marketplace/). Also seems to be the most common case for @TBBle. It seems that to support these use cases, we would want a pair of command-line switches, e.g. `helm install --global-registry-host=my.registry.local --global-registry-namespace=project23`.

https://github.com/pivotal/image-relocation and its [users](https://github.com/pivotal/image-relocation#where-is-this-repository-used) allow images to be relocated using a registry prefix which can be just the registry, the registry and the "namespace", or more (for registries like GCR which support multiple path components).

Some other tools, such as https://github.com/cnabio/cnab-to-oci, relocate images to a single repository. This has a downside that relocatable image references become pretty inscrutable to humans, but it is handy for ECR (which requires each repository to be created before an image can be pushed).

**Comment by jhvhs** [comment]
I have updated the proposal to reflect all of the feedback I've gotten so far.

**Comment by TBBle** [comment]
Sorry, just catching up. To clarify, as I think this was unclear: The use-case we have worked with here is that we globally override the OCI registry for _our product's images_ to point to a registry visible in the DMZ, but not for images pulled from public sources, as the public sources are directly visible from the DMZ.

Although I imagine our deployment teams might _also_ want to maintain a private mirror registry of public images for validation and locking floating tags (or just to lock down the DMZ harder), but they would (hopefully) use a _different_ local registry than the one we would be mirroring our product's images to.

So we would currently override just our own chart's images, i.e. "everything from internal_project_registry" but I can see the use-case for overriding "everything from docker.io", and "everything from quay.io".

I definitely agree with @glyn that the SHA256 of the image would _not_ change when mirroring, and I would not expect non-floating tags to give different results on any mirror either.

https://stevelasker.blog/2020/02/17/registry-namespace-repo-names/ is an interesting read. I'll pass it around my team, as it might help align some language and thinking in our registry layout, as "namespace" is new to me here, and helps solve a terminology conflict.

**Comment by jhvhs** [comment]
@TBBle To cover your use case, it's possible to enhance the `applyTo` setting. In this case, rather than using a single string, we could use a key/value pair with one or more of the following keys:
- `imageSelector` - (sorry, I can't come up with a better name right now) the value described above in the proposal for selecting images by name/chart name. Setting this to `*.*` would not make any sense when other keys are present.
- `registry` - images with the matching registry, where unspecified registry would be expanded to `docker.io`.
- `namespace` - would match all the images with the specified namespace.

Example:
```yaml
locations:
- registry: project-254.delhi.internal
  applyTo: # Applies only to the images in the current chart hosted on project-254.quebec.internal
  - imageSelector: '*'
    registry: project-254.quebec.internal
```

What do you think?

**Comment by TBBle** [comment]
At this point, I understand that for this idea to be advanced, it will need a Helm Improvement Proposal. See [HIP-0001](https://github.com/helm/community/blob/master/hips/hip-0001.md) for details on the process.

**Comment by lindhe** [comment]
Plz bot no stale :)

**Comment by cndoit18** [comment]
Plz bot no stale again :)

**Comment by lindhe** [comment]
Before I forget it, I want to propose that we name the new file `Images.yaml` instead of `images.yaml`. When referring to the objects in the templates, they are called `.Chart` and `.Values` respectively. And while `Chart.yaml` already follows the same convention, it has always bugged me that `values.yaml` has a lowercase v. :)

**Comment by technosophos** [comment]
The reason `values.yaml` is lowercase is because it is (and was supposed to be) conventional, but not required. Chart.yaml _must_ be named `Chart.yaml`. So if the file will have special significance to Helm (e.g. the name is hardcoded), then it should be `Images.yaml`. If it is merely conventional, then it should be `images.yaml`.

**Comment by bweston92** [comment]
@TBBle does a HIP need creating still? If not can we link it here.
